src/Microsoft.OData.Client/DataServiceExtensions.cs

```diff
@@ -0,0 +1,173 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="DataServiceExtensions.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.Client
+{
+    #region Namespaces
+
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Linq.Expressions;
+    using System.Reflection;
+
+    #endregion Namespaces
+
+    public static class DataServiceExtensions
+    {
+        /// <summary>
+        /// Returns the distinct count of elements in a sequence after applying the projection function to each element.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source"/></typeparam>
+        /// <typeparam name="TTarget">The type returned by the projection function represented in <paramref name="selector"/>.</typeparam>
+        /// <param name="source">A sequence of values of type <typeparamref name="TSource"/>.</param>
+        /// <param name="selector">A projection function to apply to each element.</param>
+        /// <returns>Distinct count of elements in a sequence after applying the projection function to each element.</returns>
+        public static int CountDistinct<TSource, TTarget>(this IQueryable<TSource> source, Expression<Func<TSource, TTarget>> selector)
+        {
+            MethodCallExpression countMethodExpr;
+
+            if (source.Provider.GetType().Equals(typeof(DataServiceQueryProvider)))

```

Comment:
>source.Provider.GetType().Equals(typeof(DataServiceQueryProvider))) [](start = 16, length = 67)

Could the provider type be the sub type of "DataServiceQueryProvider"?

Reply:
@xuzhg Thanks to your observation here. As a result of looking at it afresh I was able to significantly reduce the code in the extension method